src/Features/Core/Portable/Completion/CompletionService.cs

```diff
@@ -249,39 +252,47 @@ public virtual TextSpan GetDefaultCompletionListSpan(SourceText text, int caretP
             string filterText)
         {
             var helper = CompletionHelper.GetHelper(document);
-            var itemsWithPatternMatch = new SegmentedList<(CompletionItem, PatternMatch?)>(items.Select(
-                item => (item, helper.GetMatch(item, filterText, includeMatchSpans: false, CultureInfo.CurrentCulture))));
+            var filterDataList = new SegmentedList<MatchResult>(items.Select(
+                item => helper.GetMatchResult(item, filterText, includeMatchSpans: false, CultureInfo.CurrentCulture)));
 
-            var builder = ImmutableArray.CreateBuilder<CompletionItem>();
-            FilterItems(helper, itemsWithPatternMatch, filterText, builder);
-            return builder.ToImmutable();
+            var builder = s_listOfMatchResultPool.Allocate();
+            try
+            {
+                FilterItems(helper, filterDataList, filterText, builder);
+                return builder.SelectAsArray(result => result.CompletionItem);
+            }
+            finally
+            {
+                // Don't call ClearAndFree, which resets the capacity to a default value.
+                builder.Clear();
+                s_listOfMatchResultPool.Free(builder);
+            }
         }
 
         internal virtual void FilterItems(
            Document document,
-           IReadOnlyList<(CompletionItem, PatternMatch?)> itemsWithPatternMatch,
+           IReadOnlyList<MatchResult> matchResults,
            string filterText,
-           IList<CompletionItem> builder)
+           IList<MatchResult> builder)
```

Comment:
We probably should deprecate the public version of this and promote this to public